python/lsst/daf/butler/core/datasets/type.py

```diff
@@ -240,25 +240,78 @@ def __repr__(self) -> str:
             extra += ", isCalibration=True"
         return f"DatasetType({self.name!r}, {self.dimensions}, {self._storageClassName}{extra})"
 
-    def __eq__(self, other: Any) -> bool:
+    def _equal_ignoring_storage_class(self, other: Any) -> bool:
+        """Check everything is equal except the storage class.
+
+        Parameters
+        ----------
+        other : Any
+            Object to check against this one.
+
+        Returns
+        -------
+        mostly : `bool`
+            Returns `True` if everything except the storage class is equal.
+        """
         if not isinstance(other, type(self)):
             return False
         if self._name != other._name:
             return False
         if self._dimensions != other._dimensions:
             return False
+        if self._isCalibration != other._isCalibration:
+            return False
+        if self._parentStorageClass is not None and other._parentStorageClass is not None:
+            return self._parentStorageClass == other._parentStorageClass
+        else:
+            return self._parentStorageClassName == other._parentStorageClassName
+
+    def __eq__(self, other: Any) -> bool:
+        mostly_equal = self._equal_ignoring_storage_class(other)
+        if not mostly_equal:
+            return False
+
+        # Be careful not to force a storage class to import the corresponding
+        # python code.
         if self._storageClass is not None and other._storageClass is not None:
             if self._storageClass != other._storageClass:
                 return False
         else:
             if self._storageClassName != other._storageClassName:
                 return False
-        if self._isCalibration != other._isCalibration:
+        return True
+
+    def is_compatible_with(self, other: "DatasetType") -> bool:
```

Comment:
Does "compatible" mean compatible for both `get` and `put`? Is it symmetrical (X.is_compatible_with(Y) === Y.is_compatible_with(X))?

Reply:
It really means `self_sc.can_convert(other_sc)` -- You have made me realize that ctrl_mpexec is going to have to try to distinguish inputs (so can convert on get to the required type for the Task) and outputs (so can convert the type used in the Task to the required type in the butler).